resources/todomvc/architecture-examples/angular/src/app/todos.service.ts

```diff
@@ -0,0 +1,62 @@
+import { Injectable } from '@angular/core';
+import { Todo } from './todo';
+
+function uuid() {
+    let uuid = "";
+    for (let i = 0; i < 32; i++) {
+        let random = (Math.random() * 16) | 0;
+        if (i === 8 || i === 12 || i === 16 || i === 20)
+            uuid += "-";
+
+        uuid += (i === 12 ? 4 : i === 16 ? (random & 3) | 8 : random).toString(16);
+    }
+    return uuid;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TodosService {
+  todos: Todo[] = [];
+
+  addItem(title: string): void {
+    const todo:Todo = {
+      id: uuid(),
+      title,
+      completed: false,
+    };
+
+    this.todos = this.todos.concat(todo);
+  }
+
+  toggleItem(todoToToggle: Todo):void {
+    this.todos = this.todos.map((todo) => todo !== todoToToggle ? todo : { ...todo, completed: !todo.completed });
+  }
+
+  removeItem(todo: Todo):void {
+    this.todos = this.todos.filter((t) => t !== todo);
+  }
+
+  updateItem(todoToSave: Todo, title: string) {
+    this.todos = this.todos.map((todo) => todo !== todoToSave ? todo : { ...todo, title });
+  }
```

Comment:
It would be good to check whether angular would work with mutations rather than generating the whole list (and of course look at the performance difference if this works). I'm under the impression that Angular works from the user inputs rather than the values, but I merely skimmed through the documentation.

If mutations work, it would be a good idea to use mutations when possible.

Reply:
will check

Reply:
interesting - I made the change to use mutation for toggle item and update item. 
Seems like there's a slightly better performance:
https://gist.github.com/flashdesignory/80970c02d90a8b29e611fe8bfb53e5ba